db/src/timelines.rs

```diff
@@ -0,0 +1,442 @@
+        assert_eq!(conn.schema, schema1);
+
+        assert_matches!(conn.datoms(), r#"
+            [[?e :db/ident :test/entid]
+             [?e :db/doc "test"]
+             [?e :db.schema/version 1]]
+        "#);
+        assert_matches!(conn.transactions(), r#"
+            [[[?e :db/ident :test/entid ?tx true]
+              [?e :db/doc "test" ?tx true]
+              [?e :db.schema/version 1 ?tx true]
+              [?tx :db/txInstant ?ms ?tx true]]]
+        "#);
+    }
+
+    
+    #[test]
+    fn test_pop_schema() {
+        let mut conn = TestConn::default();
+        conn.sanitized_partition_map();
+
+        let t = r#"
+            [{:db/id "e" :db/ident :test/one :db/valueType :db.type/long :db/cardinality :db.cardinality/one}
+             {:db/id "f" :db/ident :test/many :db/valueType :db.type/long :db/cardinality :db.cardinality/many}]
+        "#;
+
+        let partition_map0 = conn.partition_map.clone();
+        let schema0 = conn.schema.clone();
+
+        let report1 = assert_transact!(conn, t);
+        let partition_map1 = conn.partition_map.clone();
+        let schema1 = conn.schema.clone();
+
+        let (new_schema, new_partition_map) = move_from_main_timeline(
+            &conn.sqlite, &conn.schema, conn.partition_map.clone(),
+            &vec![report1.tx_id], 1).expect("moved single tx");
+        update_conn(&mut conn, &new_schema, &new_partition_map);
+
+        assert_matches!(conn.datoms(), "[]");
+        assert_matches!(conn.transactions(), "[]");
+        assert_eq!(conn.partition_map, partition_map0);
+        assert_eq!(conn.schema, schema0);
+
+        let report2 = assert_transact!(conn, t);
+        let partition_map2 = conn.partition_map.clone();
+        let schema2 = conn.schema.clone();
+
+        assert_eq!(report1.tx_id, report2.tx_id);
+        assert_eq!(partition_map1, partition_map2);
+        assert_eq!(schema1, schema2);
+
+        assert_matches!(conn.datoms(), r#"
+            [[?e1 :db/ident :test/one]
+             [?e1 :db/valueType :db.type/long]
+             [?e1 :db/cardinality :db.cardinality/one]
+             [?e2 :db/ident :test/many]
+             [?e2 :db/valueType :db.type/long]
+             [?e2 :db/cardinality :db.cardinality/many]]
+        "#);
+        assert_matches!(conn.transactions(), r#"
+            [[[?e1 :db/ident :test/one ?tx1 true]
+             [?e1 :db/valueType :db.type/long ?tx1 true]
+             [?e1 :db/cardinality :db.cardinality/one ?tx1 true]
+             [?e2 :db/ident :test/many ?tx1 true]
+             [?e2 :db/valueType :db.type/long ?tx1 true]
+             [?e2 :db/cardinality :db.cardinality/many ?tx1 true]
+             [?tx1 :db/txInstant ?ms ?tx1 true]]]
+        "#);
+    }
+
+    #[test]
+    fn test_pop_in_sequence() {
+        let mut conn = TestConn::default();
+        conn.sanitized_partition_map();
+
+        let partition_map_after_bootstrap = conn.partition_map.clone();
+
+        assert_eq!((65536..65538),

```

Comment:
It's a shame to have to do this allocation in these tests.  I have a plan to improve this, however: our matcher is rich, and knows what it matched.  We just need to grow `assert_matches!` to _also_ match the matcher, so that we can feed the tempids into the `assert_matches!` and verify that we really got what we expected.  Then it's all symbolic :)